### PR TITLE
231 amarna ci issue

### PR DESCRIPTION
**Pull request by shramee**
## Usage related changes

None

## Development related changes

Fixes Amarna CI test issues.

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [ ] Documented the changes
-   [x] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [ ] Linked issues which this PR resolves
-   [ ] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example):
    -   < EXAMPLE_REPO_PR_URL > 
    -   [ ] Modified `test.sh` to use my example repo branch
    -   [ ] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [ ] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

**Comment by shramee** [comment]
Continues #295

**Comment by shramee** [comment]
@FabijanC Ready to have a look now.

File: src/external-server/docker-amarna.ts
Comment (FabijanC) [inline src/external-server/docker-amarna.ts:38]: The `pullImage` method was made protected but isn't used, although this looks like a use case for it.

**Review by FabijanC — CHANGES_REQUESTED** [review]
While reviewing this, I realized we've probably been adding Amarna support the wrong way. 

The idea of the `DockerServer` abstract class is to support using Docker containers through a server. That is useful if the container needs to do time-consuming things before doing the actual useful work (e.g. importing the StarkNet compiler before compiling). Also, that is applicable only if Hardhat is expected to use the Docker server multiple times while the Hardhat process is alive. That is why `DockerServer` wasn't really suitable for `docker run -it` operations.

It wasn't always like that; we used to rely on `hardhat-docker` (an official plugin for Docker interaction in Hardhat), and that might just be what we'd need here.

I would say the e2e test is ok (once my suggested change is implemented), so that part can stay.

Let me know if anything needs more clarification or if we should discuss this on a call.

File: test/general-tests/amarna/check.ts
Comment (shramee) [reply test/general-tests/amarna/check.ts:24]: So we have a better visibility of what's going on and if it fails.

File: src/external-server/docker-amarna.ts
Comment (shramee) [reply src/external-server/docker-amarna.ts:38]: Aye, but it doesn't do it in sync, so we can't use it :(

Comment (FabijanC) [reply src/external-server/docker-amarna.ts:38]: I don't understand, why `await this.pullImage(...)` be used?

**Comment by FabijanC** [comment]
If it's ready for another round of review, I can do it, but tomorrow.

File: src/external-server/docker-amarna.ts
Comment (FabijanC) [inline src/external-server/docker-amarna.ts:17]: I'd also use slicing to remove `"0."` from the name, as we do in `DockerServer`:
```js
Math.random().toString().slice(2);
```

File: src/external-server/docker-amarna.ts
Comment (FabijanC) [inline src/external-server/docker-amarna.ts:53]: Why not log stdout to stdout and stderr to stderr?

**Review by FabijanC — CHANGES_REQUESTED** [review]
It should be documented that this feature is only available if the user has docker.

File: src/external-server/docker-amarna.ts
Comment (FabijanC) [inline src/external-server/docker-amarna.ts:49]: Why doesn't it need to bind other paths as well? Like source paths?

**Comment by shramee** [comment]
@FabijanC It's ready to have a look 😉 but not sure I understand **source paths**.

**Review by FabijanC — CHANGES_REQUESTED** [review]
@shramee And let's not forget to update the docs to say that this feature is only available if the user has Docker installed.

File: src/external-server/docker-amarna.ts
Comment (FabijanC) [reply src/external-server/docker-amarna.ts]: Resolving this conversation, continued in https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/302#discussion_r1090316707

**Comment by shramee** [comment]
Added the docs and Cairo paths future proofing 😉

**Review by FabijanC — CHANGES_REQUESTED** [review]
Should have tried this long before, but this is what I got after running `npx hardhat amarna`:
```
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
exec /usr/local/bin/amarna: exec format error
```

**Comment by FabijanC** [comment]
Also, linter is reporting an unused-import error

**Comment by FabijanC** [comment]
@shramee The warning I mentioned in [this comment](https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/302#pullrequestreview-1278754963) - that's the only thing that gets printed and no output file is generated. That's the same thing that happened on CircleCI, the image platform was incompatible. I would expect it to fail in that case, but looks like it didn't.

Also, did we see this warning in the CircleCI logs? If not, why?

**Comment by shramee** [comment]
The error for platform shouldn't affect the functioning, however I don't think I've seen the exec error. Let me know if you see that again.

**Comment by shramee** [comment]
I amended the last commit to remove `out.sarif` file. You might need to pull again...

**Comment by FabijanC** [comment]
> The error for platform shouldn't affect the functioning, however I don't think I've seen the exec error. Let me know if you see that again.

It shouldn't but it does. For me it just prints the warning (it's not even an error, it exits with code 0) and doesn't produce any file outputl

**Comment by shramee** [comment]
I think I need to deploy a multi arch image... To support both arm/v8 and amd64

**Comment by shramee** [comment]
Also, I think that `exec` error might be the one halting the everything, do we have any more info about it?

**Comment by shramee** [comment]
Yeep, I think we need multi-arch 💪

**Comment by FabijanC** [comment]
> Yeep, I think we need multi-arch muscle

Will you be able to take care of the building?

**Comment by shramee** [comment]
Aye ser, leave it to me.
But might need a while as I'm also doing a HackerBuilding project. That okay?

**Comment by FabijanC** [comment]
> Aye ser, leave it to me. But might need a while as I'm also doing a HackerBuilding project. That okay?

Ok! But let's try to resolve it at least next week then. We might then temporarily disable the Amarna test until this PR is merged (in which case this PR could then also enable back the disabled test).

**Comment by shramee** [comment]
Multi arch image is deployed :fire:

**Review by FabijanC — CHANGES_REQUESTED** [review]
Alright, it's working now with the multiarch image! I ran it on my computer and it generated `out.sarif`. I think it would be useful to print to the user where the output file has been generated.

**Comment by FabijanC** [comment]
@shramee As announced, due to temporarily disabling the Amarna test on `master`, this PR now has a conflict with the target branch. You can resolve it together with addressing the remaining change requests.

**Comment by shramee** [comment]
Woop Woop! Awesome news!
Aye indicating the generated file path sounds like an awesome idea. 🥳

**Comment by FabijanC** [comment]
@shramee Could we have this merged within the next hour or so? Planning to release a new version.

**Comment by shramee** [comment]
Sorry, I was at Jerusalem today, just seen now.

File: src/external-server/docker-amarna.ts
Comment (FabijanC) [inline src/external-server/docker-amarna.ts:102]: Why does this depend on `useShell`? What exactly is the purpose of `useShell`?

**Review by FabijanC — COMMENTED** [review]
Btw, it turns out external contributors can trigger our CircleCI workflow. I enabled that feature yesterday, so your last two commits have green checks indicating all tests have passed, including the Amarna test: [test log link](https://app.circleci.com/pipelines/github/Shard-Labs/starknet-hardhat-plugin/1587/workflows/75ea4e99-6248-4af6-b4d5-917e7852132f/jobs/7029?invite=true#step-111-473).

I just kindly ask you not to use it recklessly in future PRs. You can put `[skip ci]` in your commit message to prevent workflow execution, which is recommendable if you are only improving docs or doing something very minor.

**Pull request merged**
